ibis/backends/flink/__init__.py

```diff
@@ -853,3 +859,142 @@ def insert(
             "is not a pandas DataFrame or is not a ibis Table."
             f"The given obj is of type {type(obj).__name__} ."
         )
+
+    def to_pyarrow(
+        self,
+        expr: ir.Expr,
+        *,
+        params: Mapping[ir.Scalar, Any] | None = None,
+        limit: int | str | None = None,
+        **kwargs: Any,
+    ) -> pa.Table:
+        from ibis.formats.pyarrow import PyArrowSchema
+
+        pyarrow_batches = self.to_pyarrow_batches(
+            expr, params=params, limit=limit, **kwargs
+        )
+
+        try:
+            pa_table = pa.Table.from_batches(pyarrow_batches)
+
+        except ValueError:
+            table_expr = expr.as_table()
+            ibis_schema = table_expr.schema()
+            arrow_schema = PyArrowSchema.from_ibis(ibis_schema)
+            pa_table = pa.Table.from_batches(pyarrow_batches, schema=arrow_schema)
+
+        return expr.__pyarrow_result__(pa_table)
+
+    def to_pyarrow_batches(
+        self,
+        expr: ir.Table,
+        *,
+        params: Mapping[ir.Scalar, Any] | None = None,
+        chunk_size: int | None = None,
+        limit: int | str | None = None,
+        **kwargs: Any,
+    ):
+        try:
+            return self.to_pyarrow_batches_w_batch_iterator(
+                expr=expr,
+                params=params,
+                chunk_size=chunk_size,
+                limit=limit,
+                **kwargs,
+            )
+
+        except ValueError:
+            # Note (mehmet): In the following, the entire result is fetched
+            # into a dataframe before converting it to arrow batches.
+            #
+            # TODO (mehmet): Verify that `to_pyarrow_batches_w_batch_iterator()`
+            # brings data into memory batch by batch, rather than all at once
+            # as being done in the following.
+
+            from ibis.formats.pyarrow import PyArrowSchema
+
+            table_expr = expr.as_table()
+            df = self.execute(table_expr, limit=limit, **kwargs)
+            # TODO (mehmet): `limit` is discarded in `execute()`. Is this intentional?
+            df = df.head(limit)
+
+            ibis_schema = table_expr.schema()
+            arrow_schema = PyArrowSchema.from_ibis(ibis_schema)
+            arrow_table = pa.Table.from_pandas(df, schema=arrow_schema)
+            return arrow_table.to_reader()
+
+    def to_pyarrow_batches_w_batch_iterator(
+        self,
+        expr: ir.Table,
+        *,
+        params: Mapping[ir.Scalar, Any] | None = None,
+        chunk_size: int | None = None,
+        limit: int | str | None = None,
+        **kwargs: Any,
+    ):
```

Comment:
Thanks, I used help from PyFlink's implementation :)

Reply:
Would it make sense to add a comment with the link to the implementation? I think we know where it is/what this is based on, but it might be helpful in case somebody needs to update it in the future if it changes in Flink (or wants to understand where this came from).

Reply:
Sure, added a comment at the top of the function pointing to PyFlink implementation.